Adding search processor for score normalization and combination

martin-gaievski opened the pull request:
### Description
Added processor that does score normalization and combination for Hybrid query results. Processor is registered via search pipeline API as search phase result processor. 

In this PR we're adding support for techniques "min-max" for normalization and "arithmetic mean" for combination. Support for weights and other techniques will be added in next PRs to keep size of a single PR minimal.

Below is example of definition for registering new processor as part of search pipeline:

```
{
    "description": "Post processor for hybrid search",
    "phase_results_processors": [
        {
            "normalization-processor": {
                "normalization": {
                    "technique": "min_max"
                },
                "combination": {
                    "technique": "arithmetic_mean"
                }
            }
        }
    ]
}
```

### Issues Resolved
https://github.com/opensearch-project/neural-search/issues/228, part of solution for https://github.com/opensearch-project/neural-search/issues/126

### Check List
- [X] New functionality includes testing.
    - [X] All tests pass
- [X] New functionality has been documented.
    - [X] New functionality has javadoc added
- [X] Commits are signed as per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

navneet1v commented on src/main/java/org/opensearch/neuralsearch/processor/ScoreCombinationTechnique.java:
```diff
@@ -0,0 +1,43 @@
+            float combinedScore = 0.0f;
+            int count = 0;
+            for (float score : scores) {
+                if (score >= 0.0) {
```
why we are ignoring the 0 scores herre?

martin-gaievski replied:
I think we discussed this approach offline, score 0.0 means there is not hit for that doc in one particular sub-query. It's typically for text based queries like "term" or "match", in such case we're boosting results of neural searches.  
I think both approaches may have their use cases, we can add "vanilla" arithmetic mean and keep this one, so user chooses which one fits better.

navneet1v commented on src/main/java/org/opensearch/neuralsearch/processor/ScoreNormalizationTechnique.java:
```diff
@@ -0,0 +1,57 @@
+        @Override
+        float normalize(ScoreNormalizationRequest request) {
+            // edge case when there is only one score and min and max scores are same
+            if (Floats.compare(request.getMaxScore(), request.getMinScore()) == 0
+                && Floats.compare(request.getMaxScore(), request.getScore()) == 0) {
+                return SINGLE_RESULT_SCORE;
+            }
```
This is really interesting use case, but I am not sure making this as 0.01f will work? reason being if that result was the most relevant result, then it will never be returned to the customer. can we check with Science team on this what is their thought.

martin-gaievski replied:
I'm actually making it 1.0, max possible in normalization.

navneet1v commented on src/main/java/org/opensearch/neuralsearch/processor/ScoreNormalizationTechnique.java:
```diff
@@ -0,0 +1,57 @@
+     */
+    @Data
+    @Builder
+    static class ScoreNormalizationRequest {
```
is this a good enough Request Object for all the different technique? We should have a base class named NormalizationRequest and then its concrete implementations MinMaxNormalizationRequest, ScalerNormalizationRequest etc.

martin-gaievski replied:
I was going to use this class as base class for DTOs, as we do have only one technique it fits all the requirements. Let me change the class name. My main idea was to have a DTO class rather than build a long train of single method arguments.

navneet1v commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessor.java:
```diff
@@ -0,0 +1,204 @@
+                .stream()
+                .filter(Objects::nonNull)
+                .findFirst();
+            if (isNotHybridQuery(maybeResult)) {
```
we should make the condition as rather than making it negation.

`if(hybridQueryResults(maybeResults))`

martin-gaievski replied:
Negation is needed to fail fast. Options are to negate in the if clause (```if (!hybridQueryResults(maybeResults))```) which is hard to read and lass error prone, or encapsulate that in the method.

navneet1v commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessor.java:
```diff
@@ -0,0 +1,204 @@
+                return;
+            }
+
+            TopDocsAndMaxScore[] topDocsAndMaxScores = getCompoundQueryTopDocsForResult(searchPhaseResult);
```
```suggestion
            TopDocsAndMaxScore[] topDocsAndMaxScores = getCompoundQueryTopDocsFromSearchPhaseResult(searchPhaseResult);
```

martin-gaievski replied:
ack

navneet1v commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessor.java:
```diff
@@ -0,0 +1,204 @@
+            CompoundTopDocs[] queryTopDocs = Arrays.stream(topDocsAndMaxScores)
+                .map(td -> td != null ? (CompoundTopDocs) td.topDocs : null)
+                .collect(Collectors.toList())
+                .toArray(CompoundTopDocs[]::new);
```
do we need to convert it to array why we cannot use list here?

martin-gaievski replied:
should be possible, let me check. Now such changes are much easier/safer as we do have decent test coverage

navneet1v commented on src/main/java/org/opensearch/neuralsearch/processor/ScoreNormalizer.java:
```diff
@@ -0,0 +1,89 @@
+     * @param queryTopDocs original query results from multiple shards and multiple sub-queries
+     * @param normalizationTechnique exact normalization method that should be applied
+     */
+    public void normalizeScores(final CompoundTopDocs[] queryTopDocs, final ScoreNormalizationTechnique normalizationTechnique) {
```
This can be a static function, as we are not storing any states here.

martin-gaievski replied:
ack

navneet1v commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessor.java:
```diff
@@ -0,0 +1,204 @@
+        TopDocsAndMaxScore[] result = new TopDocsAndMaxScore[results.getAtomicArray().length()];
+        int idx = 0;
+        for (Result shardResult : preShardResultList) {
+            if (shardResult == null) {
+                idx++;
```
one thing I am not understanding here that if results are null we are moving the idx via idx++, same at line 178 and 179, why we are doing that?

martin-gaievski replied:
it's more like temporary int index for iterating over collection in loop. I'll rewrite it to explicit for loop to make it more clear

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/ScoreCombiner.java:
```diff
@@ -0,0 +1,97 @@
+            );
+            // we're merging docs with normalized and combined scores. we need to have only maxHits results
+            for (int docId : normalizedScoresPerDoc.keySet()) {
+                pq.add(docId);
```
```suggestion
                pq.addAll(normalizedScoresPerDoc.keySet());
```

martin-gaievski replied:
ack

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/ScoreCombiner.java:
```diff
@@ -0,0 +1,97 @@
+            }
+            compoundQueryTopDocs.scoreDocs = finalScoreDocs;
+            compoundQueryTopDocs.totalHits = new TotalHits(maxHits, totalHits);
+            log.info(String.format(Locale.ROOT, "update top docs maxScore, updated value %f", maxScore));
```
```suggestion
            log.info("update top docs maxScore, updated value %", maxScore);
```

heemin32 replied:
Wouldn't it be too much to log this info for every single search?

martin-gaievski replied:
yup, removed, it's an artifact of some early stage debugging

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/ScoreCombiner.java:
```diff
@@ -0,0 +1,97 @@
+            ScoreDoc[] finalScoreDocs = new ScoreDoc[maxHits];
+            float maxScore = combinedNormalizedScoresByDocId.get(pq.peek());
+
+            for (int j = 0; j < maxHits; j++) {
```
maxHits will be always smaller than pq.size()?
From the code itself, it is not clear.

martin-gaievski replied:
it should, but let me add the check

navneet1v commented on src/main/java/org/opensearch/neuralsearch/processor/ScoreNormalizer.java:
```diff
@@ -0,0 +1,89 @@
+ * Abstracts normalization of scores in query search results.
+ */
+@Log4j2
+public class ScoreNormalizer {
```
I am not super convinced in having this class. I think we should move the actual min,max score calculation and updating the scores to ScoreNormalizationTechnique enum class. Over there too, I would like to split the code in smaller functions like getMinScore, get MaxScore, update Score for TopDocs so that they can be used across Normalization Technique.


The current code is very specific to Min/Max normalization and doesn't seems like it can be extended easily.

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/ScoreNormalizationTechnique.java:
```diff
@@ -0,0 +1,57 @@
+     */
+    MIN_MAX {
+        @Override
+        float normalize(ScoreNormalizationRequest request) {
```
```suggestion
        public float normalize(final float score, final float minScore, final float maxScore) {
```

martin-gaievski replied:
I actually try to avoid such long list of arguments, as enum abstract method will be implemented by each technique, and most likely every one of it will need different set of params. So instead of adding arguments and pass some mocked values and prefer using DTO, it's more resilient to such changes

heemin32 replied:
Sharing single DTO and using different fields for each technique is not a good design pattern because you don't know which values in the DTO is being used for each of technique. If more technique is added, it will get more complex to understand which parameter is for which techniques.

martin-gaievski replied:
I have doubts about that statement, I think in some cases it's a good pattern. I had in mind the idea of sharing same interface across all techniques and for that having single interface is required. In such case approach with each param being an argument has the same cons that you've mentioned - you need to set say 7-8 arguments but not all of them is needed for every technique, and on top of that code is longer which makes it harder to read, and every addition or deletion of an argument requires a change in interface and every implementing class. 
My idea was that the DTO is a base class, and each technique extends it if needed and it knows what params are required. Just in this PR it's a single technique for each of sub-processor so it's not that obvious. Also after Navneet's comments I will probably redesign processor and normalizer as current version is very lean to a single method of min-max.

heemin32 replied:
Sharing same interface across multiple techniques could be a better approach. For example
```
public interface ScoreNormalization {
  float normalize(final float score);
}

public class MinMaxNormalization implements ScoreNormalization {
  private final float minScore;
  private final float maxScore;
  
  public float normalize(final float score) {
    //implementation
  }
  
  // Builder of MinMaxNormalization
}
```

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/ScoreNormalizationTechnique.java:
```diff
@@ -0,0 +1,57 @@
+/**
+ * Collection of techniques for score normalization
+ */
+public enum ScoreNormalizationTechnique {
```
Nit. Could be better to move enums to its own package?

martin-gaievski replied:
I would left it for later development, for now ```processor``` package looks enough

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/ScoreNormalizationTechnique.java:
```diff
@@ -0,0 +1,57 @@
+     */
+    @Data
+    @Builder
+    static class ScoreNormalizationRequest {
```
Can we remove this class or move it out of enum class at least?

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/plugin/NeuralSearch.java:
```diff
@@ -90,9 +97,18 @@ public Map<String, Processor.Factory> getProcessors(Processor.Parameters paramet
+            log.info("Registering hybrid query phase searcher");
             return Optional.of(new HybridQueryPhaseSearcher());
         }
+        log.info("Not registering hybrid query phase searcher because feature flag is disabled");
```
ack

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/plugin/NeuralSearch.java:
```diff
@@ -90,9 +97,18 @@ public Map<String, Processor.Factory> getProcessors(Processor.Parameters paramet
     @Override
     public Optional<QueryPhaseSearcher> getQueryPhaseSearcher() {
         if (FeatureFlags.isEnabled(NEURAL_SEARCH_HYBRID_SEARCH_ENABLED)) {
+            log.info("Registering hybrid query phase searcher");
```
ack

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/ScoreNormalizer.java:
```diff
@@ -0,0 +1,89 @@
+public class ScoreNormalizer {
+
+    /**
+     * Performs score normalization based on input combination technique. Mutates input object by updating normalized scores.
```
```suggestion
     * Performs score normalization based on input normalization technique. Mutates input object by updating normalized scores.
```

martin-gaievski replied:
ack

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/ScoreNormalizer.java:
```diff
@@ -0,0 +1,89 @@
+     */
+    public void normalizeScores(final CompoundTopDocs[] queryTopDocs, final ScoreNormalizationTechnique normalizationTechnique) {
+        Optional<CompoundTopDocs> maybeCompoundQuery = Arrays.stream(queryTopDocs)
+            .filter(topDocs -> Objects.nonNull(topDocs) && !topDocs.getCompoundTopDocs().isEmpty())
```
```suggestion
            .filter(topDocs -> Objects.nonNull(topDocs))
            .filter(topDocs -> topDocs.getCompoundTopDocs().isEmpty() == false)
```

martin-gaievski replied:
ack

martin-gaievski replied:
ack

navneet1v commented on src/main/java/org/opensearch/neuralsearch/processor/ScoreCombiner.java:
```diff
@@ -0,0 +1,97 @@
+            Map<Integer, Float> combinedNormalizedScoresByDocId = normalizedScoresPerDoc.entrySet()
+                .stream()
+                .collect(Collectors.toMap(Map.Entry::getKey, entry -> combinationTechnique.combine(entry.getValue())));
+            // create priority queue, make it max heap by the score
+            PriorityQueue<Integer> pq = new PriorityQueue<>(
+                (a, b) -> Float.compare(combinedNormalizedScoresByDocId.get(b), combinedNormalizedScoresByDocId.get(a))
+            );
+            // we're merging docs with normalized and combined scores. we need to have only maxHits results
+            for (int docId : normalizedScoresPerDoc.keySet()) {
+                pq.add(docId);
+            }
```
can we use HitQueue interface provided by lucene here? rather than using our own PQ ?

martin-gaievski replied:
I'm not sure we'll benefit from that class. It's simplifies  work with heap in case elements are updated after being added to the heap, but it's not our case. Other than that it's same standard java PriorityQueue. Also it's designed for ScoreDoc instances, but we're operating with scores.

navneet1v commented on src/main/java/org/opensearch/neuralsearch/processor/ScoreCombiner.java:
```diff
@@ -0,0 +1,97 @@
+                int docId = pq.poll();
+                finalScoreDocs[j] = new ScoreDoc(docId, combinedNormalizedScoresByDocId.get(docId), shardId);
+            }
+            compoundQueryTopDocs.scoreDocs = finalScoreDocs;
+            compoundQueryTopDocs.totalHits = new TotalHits(maxHits, totalHits);
```
shouldn't there be setter for setting these attributes? if yes then lets use them

martin-gaievski replied:
unfortunately there are no setter/getter methods

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessor.java:
```diff
@@ -0,0 +1,204 @@
+ */
+@Log4j2
+public class NormalizationProcessor implements SearchPhaseResultsProcessor {
+    public static final String TYPE = "normalization-processor";
+    public static final String NORMALIZATION_CLAUSE = "normalization";
+    public static final String COMBINATION_CLAUSE = "combination";
+    public static final String TECHNIQUE = "technique";
```
it's used from some classes outside the package, for instance [NormalizationProcessorFactory](https://github.com/opensearch-project/neural-search/blob/feature/normalization-injector-processor/src/main/java/org/opensearch/neuralsearch/processor/factory/NormalizationProcessorFactory.java)

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessor.java:
```diff
@@ -0,0 +1,204 @@
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import joptsimple.internal.Strings;
```
ack

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/ScoreNormalizer.java:
```diff
@@ -0,0 +1,89 @@
+                    float originalScore = scoreDoc.score;
+                    scoreDoc.score = normalizationTechnique.normalize(normalizationRequest);
+                    log.info(
+                        String.format(
```
```suggestion
                        log.info("update dock[%] score, original value %, updated value %", scoreDoc.doc, originalScore, scoreDoc.score)
```

martin-gaievski replied:
ack

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/ScoreNormalizer.java:
```diff
@@ -0,0 +1,89 @@
+                            .build();
+                    float originalScore = scoreDoc.score;
+                    scoreDoc.score = normalizationTechnique.normalize(normalizationRequest);
+                    log.info(
```
Could be better not logging this info?

martin-gaievski replied:
ack, leftover after debugging

martin-gaievski replied:
yes, removing

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessor.java:
```diff
@@ -0,0 +1,204 @@
+    }
+
+    protected void validateParameters(final String normalizationTechniqueName, final String combinationTechniqueName) {
+        if (Strings.isNullOrEmpty(normalizationTechniqueName)) {
```
ack

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/ScoreNormalizer.java:
```diff
@@ -0,0 +1,89 @@
+            List<TopDocs> topDocsPerSubQuery = compoundQueryTopDocs.getCompoundTopDocs();
+            for (int j = 0; j < topDocsPerSubQuery.size(); j++) {
+                TopDocs subQueryTopDoc = topDocsPerSubQuery.get(j);
+                for (ScoreDoc scoreDoc : subQueryTopDoc.scoreDocs) {
```
```suggestion
                subQueryTopDoc.scoreDocs.stream()
                .forEach(scoreDoc -> scoreDoc.score = normalizationTechnique.normalize(scoreDoc.score, minMaxScores[j][0], minMaxScores[j][1]))
```

martin-gaievski replied:
already refactored to a list

martin-gaievski replied:
ack

navneet1v commented on src/main/java/org/opensearch/neuralsearch/processor/ScoreCombiner.java:
```diff
@@ -0,0 +1,97 @@
+     * @param combinationTechnique exact combination method that should be applied
+     * @return list of max combined scores for each shard
+     */
+    public List<Float> combineScores(final CompoundTopDocs[] queryTopDocs, final ScoreCombinationTechnique combinationTechnique) {
```
Couple of comments on this function:
1. This function is long and has a lot of moving pieces. Lets break it into smaller pieces
2. Let's try to add inline comments on what we are trying to do. Like when scores are getting updated, when the scores are getting combined etc. This will ensure better readability.
3. I would like to add comments in a step by step format like we say we need to do 4 things and then mention inline when those steps are performed,

martin-gaievski replied:
ack

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessor.java:
```diff
@@ -0,0 +1,204 @@
+            throw new IllegalArgumentException("combination technique cannot be empty");
+        }
+        if (!EnumUtils.isValidEnum(ScoreNormalizationTechnique.class, normalizationTechniqueName)) {
+            log.error(String.format(Locale.ROOT, "provided normalization technique [%s] is not supported", normalizationTechniqueName));
```
I put customer provided values only to log message, and exception has only general info that "value is invalid". This is to minimize security risks of rendering user input back, as this may be a malicious content

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessor.java:
```diff
@@ -0,0 +1,204 @@
+    public NormalizationProcessor(
+        final String tag,
+        final String description,
+        final String normalizationTechnique,
+        final String combinationTechnique
```
in this case we'll need to move validation and conversion logic to factory. I'm not sure if that's better than having that responsibility in the processor.

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessor.java:
```diff
@@ -0,0 +1,209 @@
+        return result;
+    }
+
+    @VisibleForTesting
+    protected <Result extends SearchPhaseResult> void updateOriginalQueryResults(
```
because it's more visibility than "private", and the only reason for that is accessing from test classes.

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessor.java:
```diff
@@ -0,0 +1,204 @@
+        final SearchPhaseResults<Result> searchPhaseResult,
+        final SearchPhaseContext searchPhaseContext
+    ) {
+        if (searchPhaseResult instanceof QueryPhaseResultConsumer) {
```
ack

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessor.java:
```diff
@@ -0,0 +1,204 @@
+        final SearchPhaseContext searchPhaseContext
+    ) {
+        if (searchPhaseResult instanceof QueryPhaseResultConsumer) {
+            QueryPhaseResultConsumer queryPhaseResultConsumer = (QueryPhaseResultConsumer) searchPhaseResult;
```
ack, will refactor it into a single method

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessor.java:
```diff
@@ -0,0 +1,204 @@
+                .collect(Collectors.toList())
+                .toArray(CompoundTopDocs[]::new);
+
+            ScoreNormalizer scoreNormalizer = new ScoreNormalizer();
```
it's a good idea. Although I'm not sure if those concepts will survive changes that Navneet suggested, he want to use static methods. Let me address his comments and then come back to this. Overall I like the idea

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessor.java:
```diff
@@ -0,0 +1,204 @@
+
+    protected void validateParameters(final String normalizationTechniqueName, final String combinationTechniqueName) {
+        if (Strings.isNullOrEmpty(normalizationTechniqueName)) {
+            throw new IllegalArgumentException("normalization technique cannot be empty");
```
ack, your comment is second, moving this logic to factory

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessor.java:
```diff
@@ -0,0 +1,204 @@
+    private boolean isNotHybridQuery(final Optional<SearchPhaseResult> maybeResult) {
+        return maybeResult.isEmpty()
+            || Objects.isNull(maybeResult.get().queryResult())
+            || !(maybeResult.get().queryResult().topDocs().topDocs instanceof CompoundTopDocs);
```
ack, will add the check

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessor.java:
```diff
@@ -0,0 +1,204 @@
+    ) {
+        List<Result> preShardResultList = results.getAtomicArray().asList();
+        for (int i = 0; i < preShardResultList.size(); i++) {
+            QuerySearchResult querySearchResult = preShardResultList.get(i).queryResult();
```
ack

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/ScoreCombinationTechnique.java:
```diff
@@ -0,0 +1,43 @@
+    ARITHMETIC_MEAN {
+
+        @Override
+        public float combine(float[] scores) {
```
ack

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/ScoreCombiner.java:
```diff
@@ -0,0 +1,97 @@
+     * @param combinationTechnique exact combination method that should be applied
+     * @return list of max combined scores for each shard
+     */
+    public List<Float> combineScores(final CompoundTopDocs[] queryTopDocs, final ScoreCombinationTechnique combinationTechnique) {
```
will try to refactor it

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/ScoreCombiner.java:
```diff
@@ -0,0 +1,97 @@
+                TopDocs topDocs = topDocsPerSubQuery.get(j);
+                int hits = 0;
+                for (ScoreDoc scoreDoc : topDocs.scoreDocs) {
+                    if (!normalizedScoresPerDoc.containsKey(scoreDoc.doc)) {
```
changed, but not sure if result is better, it's mainly because it's not possible to generate array of floats and fill it in one call, and putIfAbsent doesn't take the functional inteface.

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java:
```diff
@@ -0,0 +1,161 @@
+        float maxScore = combinedNormalizedScoresByDocId.get(scoreQueue.peek());
+
+        // - update query search results with normalized scores
+        updateQueryTopDocsWithCombinedScores(compoundQueryTopDocs, topDocsPerSubQuery, combinedNormalizedScoresByDocId, scoreQueue);
```
This method is doing two things: 1. updating queryTopDocs, 2. finding and returning maxScore which violate single responsibility principal. Would there be a way to separate the two responsibility out without compromising efficiency?

martin-gaievski replied:
Agree, it does both, but the alternative is not very efficient, we'll need to iterate over all updated docs and get max scores. This is comparing to approach that I took, it  reuses internals of combiner as we have that max score for update of total hits, and this is without extra result scan.

heemin32 replied:
I will leave the decision to you but the rule of thumb that I know is not to increase code complexity for non-critical optimization.

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java:
```diff
@@ -0,0 +1,161 @@
+
+        int shardId = compoundQueryTopDocs.scoreDocs[0].shardIndex;
+        for (int j = 0; j < maxHits; j++) {
+            int docId = scoreQueue.poll();
```
Is `maxHits` smaller than or equal to scoreQueue size?

martin-gaievski replied:
should be, but let me add isEmpty as second condition for loop

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java:
```diff
@@ -0,0 +1,161 @@
+
+        int shardId = compoundQueryTopDocs.scoreDocs[0].shardIndex;
+        for (int j = 0; j < maxHits; j++) {
+            int docId = scoreQueue.poll();
```
We should consider to use sorted list instead of priority queue. The time complexity of the `poll()` operation is `log(n)`

martin-gaievski replied:
I'm not sure time complexity is entirely correct, for ```poll()``` it's ```O(1)```. ```PriorityQueue``` and it's subclasses are used all over OpenSearch and Lucene in similar cases.

heemin32 replied:
It depends on how the priority queue is implemented. However, there is a high chance that current implementation is O(logn). https://docs.oracle.com/en/java/javase/13/docs/api/java.base/java/util/PriorityQueue.html

martin-gaievski replied:
Looks like there are no clear answer, some sources say it's O(1) and some that it's O(logn). It's not super critical section, I'll try to rewrite it with a sorted list, otherwise we can leave it for performance test time.

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java:
```diff
@@ -0,0 +1,161 @@
+        for (int j = 0; j < topDocsPerSubQuery.size(); j++) {
+            TopDocs topDocs = topDocsPerSubQuery.get(j);
+            for (ScoreDoc scoreDoc : topDocs.scoreDocs) {
+                normalizedScoresPerDoc.putIfAbsent(scoreDoc.doc, normalizedScoresPerDoc.computeIfAbsent(scoreDoc.doc, key -> {
```
```suggestion
                normalizedScoresPerDoc.computeIfAbsent(scoreDoc.doc, key -> {
```

martin-gaievski replied:
ack, oops, over engineered here a bit

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java:
```diff
@@ -0,0 +1,161 @@
+    }
+
+    private Map<Integer, Float> combineScoresAndGetCombinedNormilizedScoresPerDocument(
+        Map<Integer, float[]> normalizedScoresPerDocument,
```
```suggestion
        final Map<Integer, float[]> normalizedScoresPerDocument,
```

martin-gaievski replied:
ack

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java:
```diff
@@ -0,0 +1,161 @@
+    }
+
+    private void updateQueryTopDocsWithCombinedScores(
+        CompoundTopDocs compoundQueryTopDocs,
```
```suggestion
        final CompoundTopDocs compoundQueryTopDocs,
```

martin-gaievski replied:
ack

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java:
```diff
@@ -0,0 +1,161 @@
+
+    private void updateQueryTopDocsWithCombinedScores(
+        CompoundTopDocs compoundQueryTopDocs,
+        List<TopDocs> topDocsPerSubQuery,
```
```suggestion
        final List<TopDocs> topDocsPerSubQuery,
```

martin-gaievski replied:
ack

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java:
```diff
@@ -0,0 +1,161 @@
+    private void updateQueryTopDocsWithCombinedScores(
+        CompoundTopDocs compoundQueryTopDocs,
+        List<TopDocs> topDocsPerSubQuery,
+        Map<Integer, Float> combinedNormalizedScoresByDocId,
```
```suggestion
        final Map<Integer, Float> combinedNormalizedScoresByDocId,
```

martin-gaievski replied:
ack

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java:
```diff
@@ -0,0 +1,161 @@
+        CompoundTopDocs compoundQueryTopDocs,
+        List<TopDocs> topDocsPerSubQuery,
+        Map<Integer, Float> combinedNormalizedScoresByDocId,
+        PriorityQueue<Integer> scoreQueue
```
```suggestion
        final PriorityQueue<Integer> scoreQueue
```

martin-gaievski replied:
ack

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java:
```diff
@@ -0,0 +1,161 @@
+        compoundQueryTopDocs.totalHits = getTotalHits(topDocsPerSubQuery, maxHits);
+    }
+
+    private int getMaxHits(List<TopDocs> topDocsPerSubQuery) {
```
```suggestion
    private int getMaxHits(final List<TopDocs> topDocsPerSubQuery) {
```

martin-gaievski replied:
ack

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java:
```diff
@@ -0,0 +1,161 @@
+        return maxHits;
+    }
+
+    private TotalHits getTotalHits(List<TopDocs> topDocsPerSubQuery, int maxHits) {
```
```suggestion
    private TotalHits getTotalHits(final List<TopDocs> topDocsPerSubQuery, final int maxHits) {
```

martin-gaievski replied:
ack

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/factory/NormalizationProcessorFactory.java:
```diff
@@ -0,0 +1,81 @@
+
+    protected void validateParameters(final String normalizationTechniqueName, final String combinationTechniqueName) {
+        if (StringUtils.isEmpty(normalizationTechniqueName)) {
+            throw new IllegalArgumentException("normalization technique cannot be empty");
```
```suggestion
            throw newConfigurationException(NormalizationProcesor.TYPE, tag, ScoreNormalizationTechnique.DEFAULT.name(), "normalization technique cannot be empty");
```

martin-gaievski replied:
nice, didn't know about this method

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationMethod.java:
```diff
@@ -0,0 +1,123 @@
+ * Abstracts normalization of scores based on min-max method
+ */
+public class MinMaxScoreNormalizationMethod implements ScoreNormalizationMethod {
+    static final float MIN_SCORE = 0.001f;
```
Can we make it either public or private?

martin-gaievski replied:
ack

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationMethod.java:
```diff
@@ -0,0 +1,123 @@
+    private static MinMaxScoreNormalizationMethod INSTANCE = new MinMaxScoreNormalizationMethod();
+
+    public static MinMaxScoreNormalizationMethod getInstance() {
+        if (Objects.isNull(INSTANCE)) {
```
Same here as mentioned above.

martin-gaievski replied:
ack

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationMethod.java:
```diff
@@ -0,0 +1,123 @@
+        if (Floats.compare(maxScore, minScore) == 0 && Floats.compare(maxScore, score) == 0) {
+            return SINGLE_RESULT_SCORE;
+        }
+        float normalizedScore = (score - minScore) / (maxScore - minScore);
```
add a unit test when maxScore and minScore is same.

martin-gaievski replied:
ack, will add in next PR

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/normalization/ScoreNormalizationTechnique.java:
```diff
@@ -0,0 +1,32 @@
+     * Min-max normalization method.
+     * nscore = (score - min_score)/(max_score - min_score)
+     */
+    MIN_MAX(MinMaxScoreNormalizationMethod.getInstance());
```
Same here as mentioned above. Could we remove the wrapping with enum?

martin-gaievski replied:
I think enum is needed, please check my comment

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/plugin/NeuralSearch.java:
```diff
@@ -90,9 +98,30 @@ public Map<String, Processor.Factory> getProcessors(Processor.Parameters paramet
     @Override
     public Optional<QueryPhaseSearcher> getQueryPhaseSearcher() {
         if (FeatureFlags.isEnabled(NEURAL_SEARCH_HYBRID_SEARCH_ENABLED)) {
+            log.info(
```
ack

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/plugin/NeuralSearch.java:
```diff
@@ -90,9 +98,30 @@ public Map<String, Processor.Factory> getProcessors(Processor.Parameters paramet
+            );
             return Optional.of(new HybridQueryPhaseSearcher());
         }
+        log.info(
```
ack

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessor.java:
```diff
@@ -0,0 +1,135 @@
+     * are set as part of class constructor
+     * @param searchPhaseResult {@link SearchPhaseResults} DTO that has query search results. Results will be mutated as part of this method execution
+     * @param searchPhaseContext {@link SearchContext}
+     * @param <Result>
```
ack

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessor.java:
```diff
@@ -0,0 +1,135 @@
+    }
+
+    private <Result extends SearchPhaseResult> List<QuerySearchResult> getQuerySearchResults(final SearchPhaseResults<Result> results) {
+        List<Result> resultsPerShard = results.getAtomicArray().asList();
```
ack

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflow.java:
```diff
@@ -0,0 +1,107 @@
+     * as we may use singleton pattern here and share one instance among processors
+     * @return instance of NormalizationProcessorWorkflow
+     */
+    public static NormalizationProcessorWorkflow create() {
```
I think you mean ```NormalizationProcessorWorkflow```, not ```NeuralSearchPlugin```. We do not control creation of ```NeuralSearchPlugin```, it's done by core. 
I think two things are mixed here, initialization with factory method and DI. In fact ```NormalizationProcessorWorkflow.create``` is a replacement for constructor, changing it to actual constructor doesn't solve mentioned problem with injection. We can create mock for ```NormalizationProcessorWorkflow``` and inject it to ```NormalizationProcessor``` for testing. I do not see benefit in moving initialization to the factory, it's main responsibility it validation and applying user provided processor config and instantiation and calling processor.

martin-gaievski replied:
ack, the suggestion is to inject workflow to the factory from the plugin class at time of factory init.

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflow.java:
```diff
@@ -0,0 +1,107 @@
+        List<CompoundTopDocs> queryTopDocs = getQueryTopDocs(querySearchResults);
+
+        // normalize
+        ScoreNormalizer scoreNormalizer = new ScoreNormalizer();
```
ack, good idea

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflow.java:
```diff
@@ -0,0 +1,107 @@
+
+    private List<CompoundTopDocs> getQueryTopDocs(List<QuerySearchResult> querySearchResults) {
+        List<CompoundTopDocs> queryTopDocs = querySearchResults.stream()
+            .filter(searchResult -> searchResult.topDocs().topDocs instanceof CompoundTopDocs)
```
this is filtered on top level, but let me add here as well

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflow.java:
```diff
@@ -0,0 +1,107 @@
+        updateOriginalQueryResults(querySearchResults, queryTopDocs, combinedMaxScores);
+    }
+
+    private List<CompoundTopDocs> getQueryTopDocs(List<QuerySearchResult> querySearchResults) {
```
ack

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflow.java:
```diff
@@ -0,0 +1,107 @@
+
+    @VisibleForTesting
+    protected void updateOriginalQueryResults(
+        List<QuerySearchResult> querySearchResults,
```
ack

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflow.java:
```diff
@@ -0,0 +1,107 @@
+    protected void updateOriginalQueryResults(
+        List<QuerySearchResult> querySearchResults,
+        final List<CompoundTopDocs> queryTopDocs,
+        List<Float> combinedMaxScores
```
ack

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflow.java:
```diff
@@ -0,0 +1,107 @@
+        final List<CompoundTopDocs> queryTopDocs,
+        List<Float> combinedMaxScores
+    ) {
+        TopDocsAndMaxScore[] topDocsAndMaxScores = new TopDocsAndMaxScore[querySearchResults.size()];
```
Let me review this method, it's part that I probably refactored the most since previous revision, so some artifact may be there

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflow.java:
```diff
@@ -0,0 +1,107 @@
+            }
+            float maxScore = updatedTopDocs.totalHits.value > 0 ? updatedTopDocs.scoreDocs[0].score : 0.0f;
+            TopDocsAndMaxScore topDocsAndMaxScore = new TopDocsAndMaxScore(updatedTopDocs, maxScore);
+            if (querySearchResult == null) {
```
ack

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflow.java:
```diff
@@ -0,0 +1,107 @@
+        }
+        for (int i = 0; i < querySearchResults.size(); i++) {
+            QuerySearchResult querySearchResult = querySearchResults.get(i);
+            CompoundTopDocs updatedTopDocs = queryTopDocs.get(i);
```
yes, one is created out of another

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/combination/ArithmeticMeanScoreCombinationMethod.java:
```diff
@@ -0,0 +1,46 @@
+    private static ArithmeticMeanScoreCombinationMethod INSTANCE = new ArithmeticMeanScoreCombinationMethod();
+
+    public static ArithmeticMeanScoreCombinationMethod getInstance() {
+        if (Objects.isNull(INSTANCE)) {
```
:) agree

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/combination/ArithmeticMeanScoreCombinationMethod.java:
```diff
@@ -0,0 +1,46 @@
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public class ArithmeticMeanScoreCombinationMethod implements ScoreCombinationMethod {
+
+    private static ArithmeticMeanScoreCombinationMethod INSTANCE = new ArithmeticMeanScoreCombinationMethod();
```
In general this may be true, but in this case the only place we use it is the enum ```ScoreCombinationTechnique``` members, I doubt mock are gonna be used there.

heemin32 replied:
Actually, I do not favor having this instance in the enum. We can discuss about it in another thread.

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/combination/ArithmeticMeanScoreCombinationMethod.java:
```diff
@@ -0,0 +1,46 @@
+                count++;
+            }
+        }
+        return combinedScore / count;
```
ack, I'll add more tests in next PRs

heemin32 replied:
Umm. I thought that divide by zero will throw an exception but I was wrong. For floating point number, it will be Infinity, -Infinity, or NaN. If those values are okay, I am fine.

martin-gaievski replied:
In such case we need to return 0.0, as this is the case when, if there are scores, we're skipping some of them and not increasing count. Let me update the method.

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombinationTechnique.java:
```diff
@@ -0,0 +1,27 @@
+ * Collection of techniques for score combination
+ */
+@AllArgsConstructor
+public enum ScoreCombinationTechnique {
```
String for technique passed by user at factory level, we can pass enum variable safer between that many layers. Plus it's a single place that encapsulates technique and method abstractions. I switched to this approach after trying technique and method or string and then method.

heemin32 replied:
Doesn't all you need is just calling combine method in the end?
Then, can we just pass `ArithmeticMeanScoreCombinationMethod` instance at factory level? I don't think passing instance is not safer than passing enum.

martin-gaievski replied:
my concern is more to keep factory light. We're going to add more methods and other configurations (like weights for sub-queries) later, and in such case factory became more responsible for kicking off calculations, while it's intended design was to parse user input and construct processor. Enum layer is taking that preparation for calculations responsibility so factory can stay light

martin-gaievski replied:
My point is that now it's a simple conversion String -> enum, and it's same as String -> method. But with planned next PR changes (and we know we'll be doing this as per [HLD RFC](https://github.com/opensearch-project/neural-search/issues/126)) it will be string + map_of_params -> method, and I'd like to keep it String -> enum. And also to keep knowledge of the actual computation method away from the factory and keep it at lower levels, factory knows just the technique, technique knows about method specifics .

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java:
```diff
@@ -0,0 +1,161 @@
+    }
+
+    private float combineShardScores(ScoreCombinationTechnique scoreCombinationTechnique, CompoundTopDocs compoundQueryTopDocs) {
+        if (Objects.isNull(compoundQueryTopDocs) || compoundQueryTopDocs.totalHits.value == 0) {
```
nope, it's set way beyond this level, otherwise this all code will be in null checks

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java:
```diff
@@ -0,0 +1,161 @@
+        }
+        List<TopDocs> topDocsPerSubQuery = compoundQueryTopDocs.getCompoundTopDocs();
+        // - create map of normalized scores results returned from the single shard
+        Map<Integer, float[]> normalizedScoresPerDoc = getNormalizedScoresPerDocument(topDocsPerSubQuery);
```
not sure what do you mean, like having a separate class and those methods protected or public. we can make them protected, it works for tests

heemin32 replied:
Having a separate class and make those method as public. This way, you can mock all those methods and unit test become simple.

heemin32 replied:
making them as `protected` does not make it mockable because they are called internally.

martin-gaievski replied:
I think ```spy()``` will provide partial mock, as we're injecting combiner to the workflow class constructor, if we create instance of combiner with ```spy()``` and then mock some methods they will be mocked

heemin32 replied:
With mokito framework, it does not work like that because mokito create proxy outside of class instead of modifying class. Therefore, If you are calling method inside class, it does not call mocked method but just call the method directly.

martin-gaievski replied:
That makes sense. I think we still can test methods individually as class doesn't have state so we passing the input and assert on output or mutated state on the input.

heemin32 replied:
Yes. You can test individual method by making it as protected. Individual method test wasn't my concern actually.
My concern was `combineScores` method in which multiple local method is called. If you are okay with writing unit test for that method, I am fine as well.

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java:
```diff
@@ -0,0 +1,161 @@
+            (a, b) -> Float.compare(combinedNormalizedScoresByDocId.get(b), combinedNormalizedScoresByDocId.get(a))
+        );
+        // we're merging docs with normalized and combined scores. we need to have only maxHits results
+        pq.addAll(normalizedScoresPerDoc.keySet());
```
ack

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java:
```diff
@@ -68,42 +71,46 @@ private float combineShardScores(
-            (a, b) -> Float.compare(combinedNormalizedScoresByDocId.get(b), combinedNormalizedScoresByDocId.get(a))
-        );
+    @VisibleForTesting
+    protected List<Integer> getPriorityQueueOfDocIds(final Map<Integer, Float> combinedNormalizedScoresByDocId) {
```
Method name needs to be updated as well. It is not PriorityQueue anymore.

martin-gaievski replied:
ack

heemin32 commented on src/test/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflowTests.java:
```diff
@@ -60,7 +60,11 @@ public void testSearchResultTypes_whenCompoundDocs_thenDoNormalizationCombinatio
+        normalizationProcessorWorkflow.execute(
+            querySearchResults,
+            ScoreNormalizationTechnique.DEFAULT,
+            new ScoreCombinationFactory().defaultCombination()
```
It should be `

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombinationFactory.java:
```diff
@@ -0,0 +1,42 @@
+     * Default combination method
+     * @return combination method that is used in case user did not provide combination technique name
+     */
+    public ScoreCombinationTechnique defaultCombination() {
```
Better to expose `public static final ScoreCombinationTechnique DEFAULT_COMBINATION_METHOD = ArithmeticMeanScoreCombinationTechnique.getInstance();` so that we don't need to create factory class every time to get default combination.

martin-gaievski replied:
agree

martin-gaievski commented on src/test/java/org/opensearch/neuralsearch/OpenSearchSecureRestTestCase.java:
```diff
@@ -150,7 +150,7 @@ protected boolean preserveIndicesUponCompletion() {
     public void deleteExternalIndices() throws IOException {
         final Response response = client().performRequest(new Request("GET", "/_cat/indices?format=json" + "&expand_wildcards=all"));
-        final XContentType xContentType = XContentType.fromMediaType(response.getEntity().getContentType());
+        final MediaType xContentType = MediaType.fromMediaType(response.getEntity().getContentType());
```
this change fixes compilation errors after recent push in core

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessor.java:
```diff
@@ -0,0 +1,129 @@
+    private final String description;
+    @VisibleForTesting
+    @Getter(AccessLevel.PACKAGE)
+    final ScoreNormalizationTechnique normalizationTechnique;
```
As there is a getter method, can we make this variable as private?

martin-gaievski replied:
ack

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:
```diff
@@ -0,0 +1,122 @@
+    private static final float MIN_SCORE = 0.001f;
+    private static final float SINGLE_RESULT_SCORE = 1.0f;
+
+    private static final MinMaxScoreNormalizationTechnique INSTANCE = new MinMaxScoreNormalizationTechnique();
```
I guess we can remove this now?
```suggestion

```

martin-gaievski replied:
ack

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:
```diff
@@ -0,0 +1,122 @@
+
+    private static final MinMaxScoreNormalizationTechnique INSTANCE = new MinMaxScoreNormalizationTechnique();
+
+    public static MinMaxScoreNormalizationTechnique getInstance() {
```
This line as well?
```suggestion

```

martin-gaievski replied:
ack

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:
```diff
@@ -0,0 +1,122 @@
+        }
+    }
+
+    private float[] getMaxScores(List<CompoundTopDocs> queryTopDocs, int numOfSubqueries) {
```
```suggestion
    private float[] getMaxScores(final List<CompoundTopDocs> queryTopDocs, final int numOfSubqueries) {
```

martin-gaievski replied:
ack

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:
```diff
@@ -0,0 +1,122 @@
+        return maxScores;
+    }
+
+    private float[] getMinScores(List<CompoundTopDocs> queryTopDocs, int numOfScores) {
```
```suggestion
    private float[] getMinScores(final List<CompoundTopDocs> queryTopDocs, final int numOfScores) {
```

martin-gaievski replied:
ack

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:
```diff
@@ -0,0 +1,122 @@
+        return minScores;
+    }
+
+    private float normalizeSingleScore(float score, float minScore, float maxScore) {
```
```suggestion
    private float normalizeSingleScore(final float score, final float minScore, final float maxScore) {
```

martin-gaievski replied:
ack

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:
```diff
@@ -0,0 +1,122 @@
+        }
+    }
+
+    private float[] getMaxScores(List<CompoundTopDocs> queryTopDocs, int numOfSubqueries) {
```
Here is some optimization opportunity. Instead of getting max and min separately, we can get it in a single loop and return float[][] which contains both min and max.

heemin32 commented on src/main/java/org/opensearch/neuralsearch/processor/normalization/ScoreNormalizer.java:
```diff
@@ -0,0 +1,29 @@
+        }
+    }
+
+    private boolean canQueryResultsBeNormalized(List<CompoundTopDocs> queryTopDocs) {
```
```suggestion
    private boolean canQueryResultsBeNormalized(final List<CompoundTopDocs> queryTopDocs) {
```

martin-gaievski replied:
ack

navneet1v commented:
@martin-gaievski all the comments which you have resolved please mark them as resolved.

navneet1v commented on src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java:
```diff
@@ -0,0 +1,172 @@
+    ) {
+        ScoreDoc[] finalScoreDocs = new ScoreDoc[maxHits];
+
+        int shardId = compoundQueryTopDocs.scoreDocs[0].shardIndex;
+        for (int j = 0; j < maxHits && j < sortedScores.size(); j++) {
+            int docId = sortedScores.get(j);
+            finalScoreDocs[j] = new ScoreDoc(docId, combinedNormalizedScoresByDocId.get(docId), shardId);
```
we are not updating the shard id? why?

martin-gaievski replied:
at this step we're working with a single shard results. It's the same shard id, so we need just to copy it to new results

navneet1v replied:
It would be useful if you can add the comments in the code, around which operation is at shard level and what is not.

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflow.java:
```diff
@@ -0,0 +1,85 @@
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * Class abstracts steps required for score normalization and combination, this includes pre-processing of income data
```
ack

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflow.java:
```diff
@@ -0,0 +1,85 @@
+
+/**
+ * Class abstracts steps required for score normalization and combination, this includes pre-processing of income data
+ * and post-processing for final results
```
ack

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessor.java:
```diff
@@ -0,0 +1,122 @@
+        final SearchPhaseResults<Result> searchPhaseResult,
+        final SearchPhaseContext searchPhaseContext
+    ) {
+        if (shouldSearchResultsBeIgnored(searchPhaseResult)) {
```
ack

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessor.java:
```diff
@@ -0,0 +1,122 @@
+            || !(maybeResult.get().queryResult().topDocs().topDocs instanceof CompoundTopDocs);
+    }
+
+    private <Result extends SearchPhaseResult> List<QuerySearchResult> getQuerySearchResults(final SearchPhaseResults<Result> results) {
```
ack

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessor.java:
```diff
@@ -0,0 +1,122 @@
+        return isNotHybridQuery(maybeResult);
+    }
+
+    private boolean isNotHybridQuery(final Optional<SearchPhaseResult> maybeResult) {
```
ack

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessor.java:
```diff
@@ -0,0 +1,122 @@
+        return description;
+    }
+
+    @Override
+    public boolean isIgnoreFailure() {
+        return true;
+    }
```
after rechecking caller code in core, yes, let's change it to false. Failing fast may help customers to identify and fix errors earlier, otherwise it will be just wrong results for hybrid query (same as processor isn't configured at all). Revert this flag in next commit

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflow.java:
```diff
@@ -0,0 +1,85 @@
+        return queryTopDocs;
+    }
+
+    @VisibleForTesting
```
ok, let me check if test can be changed, currently I'm using this function to check if results got updated.

martin-gaievski replied:
reworked test, now I'm asserting query results if scores were processed correctly, made method private and removed that annotation

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflow.java:
```diff
@@ -0,0 +1,85 @@
+    private List<CompoundTopDocs> getQueryTopDocs(final List<QuerySearchResult> querySearchResults) {
+        List<CompoundTopDocs> queryTopDocs = querySearchResults.stream()
+            .filter(searchResult -> Objects.nonNull(searchResult.topDocs()))
+            .filter(searchResult -> searchResult.topDocs().topDocs instanceof CompoundTopDocs)
```
it's more a safeguard, we're designing parts of being independent so we can reuse them later in other places

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflow.java:
```diff
@@ -0,0 +1,85 @@
+        updateOriginalQueryResults(querySearchResults, queryTopDocs, combinedMaxScores);
+    }
+
+    private List<CompoundTopDocs> getQueryTopDocs(final List<QuerySearchResult> querySearchResults) {
```
ack

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/factory/NormalizationProcessorFactory.java:
```diff
@@ -0,0 +1,75 @@
+        );
+        ScoreNormalizationTechnique normalizationTechnique = ScoreNormalizationFactory.DEFAULT_METHOD;
+        if (Objects.nonNull(normalizationClause)) {
+            String combinationTechnique = (String) normalizationClause.getOrDefault(NormalizationProcessor.TECHNIQUE, "");
```
ack

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessor.java:
```diff
@@ -0,0 +1,122 @@
+@AllArgsConstructor
+public class NormalizationProcessor implements SearchPhaseResultsProcessor {
+    public static final String TYPE = "normalization-processor";
+    public static final String NORMALIZATION_CLAUSE = "normalization";
+    public static final String COMBINATION_CLAUSE = "combination";
+    public static final String TECHNIQUE = "technique";
```
ack

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:
```diff
@@ -0,0 +1,116 @@
+ */
+public class MinMaxScoreNormalizationTechnique implements ScoreNormalizationTechnique {
+
+    public static final String TECHNIQUE_NAME = "min_max";
```
only for tests, let me change it, we can have test specific variables in test classes

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:
```diff
@@ -0,0 +1,116 @@
+     */
+    @Override
+    public void normalize(final List<CompoundTopDocs> queryTopDocs) {
+        Optional<CompoundTopDocs> optionalCompoundTopDocs = queryTopDocs.stream()
+            .filter(Objects::nonNull)
+            .filter(topDocs -> topDocs.getCompoundTopDocs().size() > 0)
+            .findAny();
```
ack

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java:
```diff
@@ -0,0 +1,172 @@
+        return combinedNormalizedScoresByDocId.get(sortedDocsIds.get(0));
+    }
+
+    @VisibleForTesting
```
ack

martin-gaievski commented on src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflow.java:
```diff
@@ -0,0 +1,85 @@
+        scoreNormalizer.normalizeScores(queryTopDocs, normalizationTechnique);
+
+        // combine
+        List<Float> combinedMaxScores = scoreCombiner.combineScores(queryTopDocs, combinationTechnique);
+
+        // post-process data
+        updateOriginalQueryResults(querySearchResults, queryTopDocs, combinedMaxScores);
```
I see what you mean, yes that assumption about the order is there. Let me check whether it's easier to provide a map with shardId as a key, or we can recalculate max scores inside updateOriginalQueryResults. It's will consume some cycles but is safer for the reasons you've mentioned

martin-gaievski replied:
I worked on this piece of code, found that we don't need to get that max scores as they are calculated later by the core, so I removed that list with max scores completely.

martin-gaievski commented:
> @martin-gaievski all the comments which you have resolved please mark them as resolved.

sure, I addressed all comments, @navneet1v  please check PR again
